File: packages/sdk-py/src/hela/rest.py
"""
Thin REST helpers. The WS flow is the hot path; these cover the
request/response corners — minting tokens from your backend, fetching
history out-of-band, publishing from a cron job.

`Hela` wraps `httpx.AsyncClient` — pass your own via `http_client=` to
plug in retries, metrics, or a shared pool.
"""

from __future__ import annotations

import contextlib
from typing import Any

import httpx

from hela._generated.rest import (
    HistoryResponse,
    PlaygroundToken,
    PublishRequest,
    PublishResponse,
    TokenRequest,
    TokenResponse,
)
from hela.errors import RateLimitedError, UnauthorizedError


class Hela:
    """
    REST client. One per base URL.

        async with Hela(base_url="https://iad.hela.dev", api_key=key) as hela:
            token = await hela.mint_token(sub="user-42", chans=[["read","chat:*"]])
            await hela.publish("chat:lobby", "hi", author="server")
    """

    def __init__(
        self,
        *,
        base_url: str,
        api_key: str | None = None,
        http_client: httpx.AsyncClient | None = None,
    ):
        self._base = base_url.rstrip("/")
        self._api_key = api_key
        self._owned_client = http_client is None
        self._client = http_client or httpx.AsyncClient(timeout=15.0)

    async def __aenter__(self) -> Hela:
        return self

    async def __aexit__(self, *exc: Any) -> None:
        await self.close()

    async def close(self) -> None:
        if self._owned_client:
            await self._client.aclose()

    # --- auth ------------------------------------------------------------

    async def mint_token(
        self,
        *,
        sub: str,
        chans: list[list[str]] | None = None,
        ttl_seconds: int = 3600,
    ) -> TokenResponse:
        """
        Ask the gateway to sign a short-lived HS256 JWT bound to this
        project. Use it on the WebSocket `token` param.
        """
        body = TokenRequest(
            sub=sub,
            chans=chans,
            ttl_seconds=ttl_seconds,
        ).model_dump(exclude_none=True)
        r = await self._post("/v1/tokens", body)
        return TokenResponse.model_validate(r)

    async def playground_token(self, *, sub: str | None = None) -> PlaygroundToken:
        """Issue a guest token for the public sandbox project."""
        body: dict[str, Any] = {}
        if sub is not None:
            body["sub"] = sub
        r = await self._post("/playground/token", body, auth=False)
        return PlaygroundToken.model_validate(r)

    # --- channels --------------------------------------------------------

    async def publish(
        self,
        channel: str,
        body: str,
        *,
        author: str | None = None,
        reply_to_id: str | None = None,
    ) -> PublishResponse:
        """
        Server-side publish. Bypasses the WS entirely — useful from
        cron jobs, background workers, or anywhere latency doesn't
        matter and you'd rather keep one fewer connection open.
        """
        req_body = PublishRequest(
            body=body,
            author=author,
            reply_to_id=reply_to_id,
        ).model_dump(exclude_none=True)
        r = await self._post(f"/v1/channels/{channel}/publish", req_body)
        return PublishResponse.model_validate(r)

    async def history(
        self,
        channel: str,
        *,
        before: str | None = None,
        limit: int = 50,
    ) -> HistoryResponse:
        """Cursor-paginated history via REST."""
        params: dict[str, Any] = {"limit": limit}
        if before is not None:
            params["before"] = before
        r = await self._get(f"/v1/channels/{channel}/history", params)
        return HistoryResponse.model_validate(r)

    # --- internal HTTP --------------------------------------------------

    def _headers(self, *, auth: bool = True) -> dict[str, str]:
        h = {"content-type": "application/json"}
        if auth and self._api_key:
            h["authorization"] = f"Bearer {self._api_key}"
        return h

    async def _post(self, path: str, body: Any, *, auth: bool = True) -> Any:
        r = await self._client.post(
            self._base + path,
            json=body,
            headers=self._headers(auth=auth),
        )
        return self._handle(r)

    async def _get(self, path: str, params: dict[str, Any]) -> Any:
        r = await self._client.get(
            self._base + path,
            params=params,
            headers=self._headers(auth=True),
        )
        return self._handle(r)

    @staticmethod
    def _handle(r: httpx.Response) -> Any:
        if r.status_code == 401:
            raise UnauthorizedError(r.text[:200])
        if r.status_code == 429:
            retry_ms = 0
            # Server is expected to send `{"retry_after_ms": N}` but we
            # tolerate missing/invalid JSON so clients still get the
            # right exception type, just without a backoff hint.
            with contextlib.suppress(Exception):
                retry_ms = int(r.json().get("retry_after_ms", 0))
            raise RateLimitedError(retry_after_ms=retry_ms)
        r.raise_for_status()
        return r.json()

File: packages/sdk-go/rest.go
package hela

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strconv"
	"strings"
	"time"
)

// REST is the server-side HTTP client. Use it to mint end-user JWTs
// from your backend, publish from a cron job, or fetch history
// out-of-band. One REST per base URL; safe for concurrent use.
//
// Bring your own http.Client if you want retries, metrics, or shared
// pooling:
//
//	rest := hela.NewREST("https://gateway-production-bfdf.up.railway.app", hela.RESTOptions{
//	    APIKey: os.Getenv("HELA_API_KEY"),
//	    HTTP:   myClient,
//	})
type REST struct {
	base   string
	apiKey string
	client *http.Client
}

// RESTOptions configures a REST instance.
type RESTOptions struct {
	APIKey string
	// HTTP lets callers plug in their own http.Client (retries, etc).
	// Optional; defaults to a 15-second-timeout stdlib client.
	HTTP *http.Client
}

// NewREST builds a client against the given base URL. Strip any
// trailing slash to keep JoinPath happy.
func NewREST(baseURL string, opts RESTOptions) *REST {
	client := opts.HTTP
	if client == nil {
		client = &http.Client{Timeout: 15 * time.Second}
	}
	return &REST{
		base:   strings.TrimRight(baseURL, "/"),
		apiKey: opts.APIKey,
		client: client,
	}
}

// MintToken asks the gateway to sign an HS256 JWT scoped to this
// project. Use it on the WebSocket `token` param.
func (r *REST) MintToken(ctx context.Context, req TokenRequest) (TokenResponse, error) {
	if req.TTLSeconds == 0 {
		req.TTLSeconds = 3600
	}
	var resp TokenResponse
	if err := r.post(ctx, "/v1/tokens", req, &resp, true); err != nil {
		return TokenResponse{}, err
	}
	return resp, nil
}

// PlaygroundToken issues a guest token for the public sandbox project.
// No API key required.
func (r *REST) PlaygroundToken(ctx context.Context, sub string) (PlaygroundToken, error) {
	body := map[string]any{}
	if sub != "" {
		body["sub"] = sub
	}
	var resp PlaygroundToken
	if err := r.post(ctx, "/playground/token", body, &resp, false); err != nil {
		return PlaygroundToken{}, err
	}
	return resp, nil
}

// Publish bypasses the WS entirely. Useful from cron jobs, background
// workers, or anywhere latency doesn't matter and you'd rather keep
// one fewer connection open.
func (r *REST) Publish(ctx context.Context, channel string, req PublishRequest) (PublishResponse, error) {
	var resp PublishResponse
	path := fmt.Sprintf("/v1/channels/%s/publish", url.PathEscape(channel))
	if err := r.post(ctx, path, req, &resp, true); err != nil {
		return PublishResponse{}, err
	}
	return resp, nil
}

// History fetches a cursor-paginated page via REST.
func (r *REST) History(ctx context.Context, channel string, req HistoryRequest) (HistoryReply, error) {
	if req.Limit == 0 {
		req.Limit = 50
	}
	params := url.Values{}
	params.Set("limit", strconv.Itoa(req.Limit))
	if req.Before != "" {
		params.Set("before", req.Before)
	}
	path := fmt.Sprintf("/v1/channels/%s/history?%s", url.PathEscape(channel), params.Encode())
	var resp HistoryReply
	if err := r.get(ctx, path, &resp); err != nil {
		return HistoryReply{}, err
	}
	return resp, nil
}

// --- internal HTTP ---------------------------------------------------

func (r *REST) post(ctx context.Context, path string, body, out any, auth bool) error {
	data, err := json.Marshal(body)
	if err != nil {
		return fmt.Errorf("hela: marshal request: %w", err)
	}
	req, err := http.NewRequestWithContext(ctx, "POST", r.base+path, bytes.NewReader(data))
	if err != nil {
		return err
	}
	r.addHeaders(req, auth)
	return r.do(req, out)
}

func (r *REST) get(ctx context.Context, path string, out any) error {
	req, err := http.NewRequestWithContext(ctx, "GET", r.base+path, nil)
	if err != nil {
		return err
	}
	r.addHeaders(req, true)
	return r.do(req, out)
}

func (r *REST) addHeaders(req *http.Request, auth bool) {
	req.Header.Set("content-type", "application/json")
	if auth && r.apiKey != "" {
		req.Header.Set("authorization", "Bearer "+r.apiKey)
	}
}

func (r *REST) do(req *http.Request, out any) error {
	resp, err := r.client.Do(req)
	if err != nil {
		return fmt.Errorf("hela: request: %w", err)
	}
	defer resp.Body.Close()

	switch resp.StatusCode {
	case 401:
		data, _ := io.ReadAll(resp.Body)
		detail := string(data)
		if len(detail) > 200 {
			detail = detail[:200]
		}
		return &UnauthorizedError{Detail: detail}
	case 429:
		data, _ := io.ReadAll(resp.Body)
		retry := 0
		var parsed struct {
			RetryAfterMs int `json:"retry_after_ms"`
		}
		// Tolerate missing or malformed body — still return the right
		// typed error, just without the retry hint.
		_ = json.Unmarshal(data, &parsed)
		retry = parsed.RetryAfterMs
		return &RateLimitedError{RetryAfterMs: retry}
	}
	if resp.StatusCode >= 400 {
		data, _ := io.ReadAll(resp.Body)
		return fmt.Errorf("hela: HTTP %d: %s: %w", resp.StatusCode, string(data), ErrHela)
	}
	if out == nil {
		return nil
	}
	return json.NewDecoder(resp.Body).Decode(out)
}
